File: Day-10: Recursion and Backtracking/1) Print all permutations of a string.java
class Solution {
    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> ans = new ArrayList<>();
        helper(nums, ans, 0);
        return ans;
    }
    public void helper(int[] nums, List<List<Integer>> ans, int id)
    {
        int n = nums.length;
        if(id == n-1)
        {
            List<Integer> temp = new ArrayList<>();
            for(int i : nums)
            {
                temp.add(i);
            }
            ans.add(temp);
            return;
        }
        for(int j = id; j < n; j++)
        {
            swap(nums, id, j);
            helper(nums, ans, id+1);
            swap(nums, id, j);
        }
    }
    public void swap(int[] nums, int i, int j)
    {
        int temp = nums[i];
        nums[i] = nums[j];
        nums[j] = temp;
    }
}

File: Day-10: Recursion and Backtracking/1) Print all permutations of a string.cpp
class Solution {
public:
    vector<vector<int>> permute(vector<int>& nums) {
        int n = nums.size();
        vector<vector<int>> ans;
        helper(0, nums, ans);
        return ans;
    }
    void helper(int id, vector<int> &nums, vector<vector<int>> &ans)
    {
        int n = nums.size();
        if(id == n)
        {
            ans.push_back(nums);
            return;
        }
        for(int i = id; i < n; i++)
        {
            swap(nums[id], nums[i]);
            helper(id+1, nums, ans);
            swap(nums[id], nums[i]);
        }
    }
};
